src/index/diskann/diskann.cc

```diff
@@ -376,48 +379,65 @@ DiskANNIndexNode<T>::Prepare(const Config& cfg) {
-        LOG_KNOWHERE_ERROR_ << "Failed to generate cache, num_nodes_to_cache(" << num_nodes_to_cache
-                            << ") is larger than 1/3 of the total data number.";
-        return false;
-    }
-    if (num_nodes_to_cache > 0) {
-        std::vector<uint32_t> node_list;
-        LOG_KNOWHERE_INFO_ << "Caching " << num_nodes_to_cache << " sample nodes around medoid(s).";
-        if (prep_conf.use_bfs_cache) {
-            auto gen_cache_expect = TryDiskANNCall<bool>([&]() -> bool {
-                pq_flash_index_->cache_bfs_levels(num_nodes_to_cache, node_list);
-                return true;
-            });
-
-            if (!gen_cache_expect.has_value()) {
-                LOG_KNOWHERE_ERROR_ << "Failed to generate bfs cache for DiskANN.";
-                return false;
-            }
-
-        } else {
-            auto gen_cache_expect = TryDiskANNCall<bool>([&]() -> bool {
-                pq_flash_index_->generate_cache_list_from_sample_queries(warmup_query_file, 15, 6, num_nodes_to_cache,
-                                                                         node_list);
-                return true;
-            });
-
-            if (!gen_cache_expect.has_value()) {
-                LOG_KNOWHERE_ERROR_ << "Failed to generate cache from sample queries for DiskANN.";
-                return false;
-            }
+    auto cached_nodes_file = diskann::get_cached_nodes_file(index_prefix_);
+    std::vector<uint32_t> node_list;
+    if (file_exists(cached_nodes_file)) {
+        LOG_KNOWHERE_INFO_ << "Reading cached nodes from file.";
+        size_t num_nodes, nodes_id_dim;
+        uint32_t* cached_nodes_ids = nullptr;
+        diskann::load_bin<uint32_t>(cached_nodes_file, cached_nodes_ids, num_nodes, nodes_id_dim);
+        node_list.assign(cached_nodes_ids, cached_nodes_ids + num_nodes);
+        if (cached_nodes_ids != nullptr) {
+            delete[] cached_nodes_ids;
         }
-        auto load_cache_expect = TryDiskANNCall<bool>([&]() -> bool {
-            pq_flash_index_->load_cache_list(node_list);
-            return true;
-        });
-
-        if (!load_cache_expect.has_value()) {
-            LOG_KNOWHERE_ERROR_ << "Failed to load cache for DiskANN.";
+    } else {
+        auto num_nodes_to_cache = GetCachedNodeNum(prep_conf.search_cache_budget_gb, pq_flash_index_->get_data_dim(),
+                                                   pq_flash_index_->get_max_degree());
+        if (num_nodes_to_cache > pq_flash_index_->get_num_points() / 3) {
+            LOG_KNOWHERE_ERROR_ << "Failed to generate cache, num_nodes_to_cache(" << num_nodes_to_cache
+                                << ") is larger than 1/3 of the total data number.";
             return false;
         }
+        if (num_nodes_to_cache > 0) {
+            LOG_KNOWHERE_INFO_ << "Caching " << num_nodes_to_cache << " sample nodes around medoid(s).";
+            std::vector<uint32_t> node_list;
```

Comment:
shadows the outter `node_list`, maybe failed to call `load_cache_list` below?

Reply:
done